model/postlude/validate_config.sail

```
@@ -282,6 +282,18 @@ function check_extension_param_constraints() -> bool = {
   valid
 }
 
+function check_physaddr_bits() -> bool = {
+  if physaddr_bits > 32 & physaddr_bits <= 34 & xlen == 32 & not(hartSupports(Ext_Sv32)) then {
+    print_endline("Sv32 is not enabled when xlen is 32 but physaddr_bits is bigger than 32");
+    false
+  }
+  else if physaddr_bits > 34 & xlen == 32 then {
+    print_endline("physaddr_bits is bigger than 34 when xlen is 32");
+    false
+  }
+  else { true }
+}
+
```

I think its better to just check whether xlen == 32 comes with either a 32 bit or 34 bit physical address nothing else. And xlen == 64 with a 56 bit physical addr.

```
function check_physaddr_bits() -> bool = {
  if xlen == 32 {
    if physaddr_bits != 34 & hartSupports(Ext_Sv32)
    	...
    	false
    if physaddr_bits != 32
    	...
    	false
  }
  else if xlen == 64 & physaddr_bits != 56 {
  	...
  	false
  }
```

---

I'm pretty sure you're allowed to have 34-bit physical addresses even without Sv32. Though it would be weird and useless so we should probably report an error even though it's technically legal.

---

> I'm pretty sure you're allowed to have 34-bit physical addresses even without Sv32. Though it would be weird and useless so we should probably report an error even though it's technically legal.

Hmm I am not familier with this part, but does this mean i can also in theory have 35 bits? Is there an upper limit? And if so, why should we support configs that are useless?

---

Yes I think you could have 35 bits. It would only be useful if you had some custom Sv33 extension or whatever that supported it. But I agree, we should error for configurations that are technically legal but very likely to be a configuration mistake, like having 34 bit physical address on a system without Sv32.

So I think it should be something like this:

```
function check_physaddr_bits() -> bool =
  physaddr_bits <= (match xlen {
     32 => (if hartSupports(Ext_Sv32) then 34 else 32),
     64 => 64,
  })
```

We could I guess technically check `Ext_SvBare` - if you don't support that then having 64 physical address bits is probably a mistake. But also I think the chance anyone ever makes a chip that doesn't support SvBare is pretty close to zero.